### PR TITLE
ShellCommandTask

### PR DESCRIPTION
**Pull request by djarecka**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary

- fixing `__getstate__` in `Result`
- fixing  execution so it can run with submitter (adding read_and_display that is not a coroutine)
- adding tests

## Checklist

- [x] All tests passing
- [x] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [x] My code follows the code style of this project 
(we are using `black`: you can `pip install pre-commit`, 
run `pre-commit install` in the `pydra` directory 
and `black` will be run automatically with each commit)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

**Comment by djarecka** [comment]
I'm trying to explore `ShellCommandTask` and add tests for it, but I'm having issue when using `Submitter` even for the simplest task, see `test_node_shelltask.py::test_shell_cmd_1a `
 
@mgxd - any idea?

**Comment by mgxd** [comment]
@djarecka I think the problem lies with
https://github.com/nipype/pydra/blob/85e7b0e625fd633a1401d3c535f9467609d4fe61/pydra/engine/task.py#L202-L206

specifically `execute()`
https://github.com/nipype/pydra/blob/85e7b0e625fd633a1401d3c535f9467609d4fe61/pydra/engine/helpers.py#L165-L168

It is sending a coroutine to an already running loop - we may want to add a synchronous subprocess call that mimics `execute()` and let submitter / workers handle multiple processes.
cc @satra

**Comment by djarecka** [comment]
@satra - Mathias helped me with the execution and it works now with submitter, so I added examples for tasks with `State` and also a simple `Workflow`. Could you please check if the tests make sense for you. I decided to have splitter over `executables` and `args`, but maybe should have also other exampels

File: pydra/engine/helpers.py [hunk 1]
```diff
@@ -158,13 +159,46 @@ def make_klass(spec):
...
+    else:
+        return rc, stdout, stderr
+
+
+def read_and_display(*cmd, strip=False, hide_display=False):
+    """Capture cmd's stdout
 
+    """
+    try:
+        process = sp.run(cmd, stdout=sp.PIPE, stderr=sp.PIPE)
```
Comment (satra) [inline pydra/engine/helpers.py]: subprocess should take in environment settings, which should be settable at the shelltask level.

File: pydra/engine/helpers.py [hunk 2]
```diff
@@ -158,13 +159,46 @@ def make_klass(spec):
...
+        )
+
+
+# run the event loop with coroutine read_and_display_async
+# or run read_and_display if a loop is already running
+def execute(cmd, strip=False):
     loop = get_open_loop()
-    rc, stdout, stderr = loop.run_until_complete(read_and_display(*cmd))
+    if loop.is_running():
+        rc, stdout, stderr = read_and_display(*cmd, strip=strip)
```
Comment (satra) [inline pydra/engine/helpers.py]: is it not possible to just await here?

File: pydra/engine/tests/test_shelltask.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+def test_shell_cmd_1_nosubm(tmpdir):
+    """ simple command, no arguments
+        no submitter
+    """
+    cmd = ["pwd"]
+    shelly = ShellCommandTask(name="shelly", executable=cmd)
+    assert shelly.cmdline == " ".join(cmd)
+    res = shelly()
+    assert res.output.stdout == str(shelly.output_dir) + "\n"
```
Comment (satra) [inline pydra/engine/tests/test_shelltask.py]:
```suggestion
    assert res.output.stdout.strip() == str(shelly.output_dir)
```

File: pydra/engine/tests/test_shelltask.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    assert res.output.stderr == ""
+
+
+@pytest.mark.parametrize("plugin", Plugins)
+def test_shell_cmd_1_strip(plugin):
+    """ simple command, no arguments
+        strip option to remove \n at the end os stdout
+    """
+    cmd = ["pwd"]
+    shelly = ShellCommandTask(name="shelly", executable=cmd, strip=True)
```
Comment (satra) [inline pydra/engine/tests/test_shelltask.py]: does strip remove just the last one?

Comment (djarecka) [reply pydra/engine/tests/test_shelltask.py]: yes, but can changed

File: pydra/engine/helpers.py (hunk 2 above)
Comment (mgxd) [reply pydra/engine/helpers.py]: not without a refactor to make `Task._run_task()` async.

**Comment by satra** [comment]
@mgxd and @djarecka - i restarted the slurm build again. do make sure that build works before merging this. it would be good to track down the loop closed error. yes, it appears stochastic, but that's what would be good to figure out.

**Comment by djarecka** [comment]
@satra - this is a different error than the loop error we sometimes have. 

here we have problems with slurm even in old tests, so it looks I broke something slurm related

**Comment by djarecka** [comment]
@satra - tests are passing so if we want, we can merge it, but there are some issues that we have to think about. For example the `cmdline` doesn't return correct command when a splitter is used

**Comment by satra** [comment]
perhaps cmdline can behave like `output_directory` provide cmdline if result is available but not otherwise.

**Pull request merged**